app/gui/view/graph-editor/src/component/node/input/widget.rs

```diff
@@ -8,54 +8,181 @@ use ensogl::data::color;
+
+
+
+// ===========
+// === FRP ===
+// ===========
 
 ensogl::define_endpoints_2! {
     Input {
-        set_current_value(Option<ImString>),
-        set_focused(bool),
+        set_metadata   (Option<Metadata>),
+        set_node_data  (NodeData),
+        set_current_value (Option<ImString>),
+        set_focused       (bool),
+        set_visible       (bool),
     }
     Output {
         value_changed(Option<ImString>),
     }
 }
 
-/// Possible widgets for a node input.
-///
-/// Currently all widget types are hardcoded. This is likely to be a temporary solution. In the
-/// future the set of widget types might be dynamic, similar to visualizations.
-#[derive(Clone, Debug, CloneRef)]
-pub enum Widget {
-    /// A widget for selecting a single value from a list of available options.
-    SingleChoice(SingleChoice),
+/// Widget metadata that needs to be fetched from the graph asynchronously.
+#[derive(Debug, Clone)]
+pub struct Metadata {
+    /// The kind of widget to use.
+    pub kind:            Kind,
+    /// The widget display mode.
+    pub display:         Display,
+    /// Dynamically fetched widget entries.
+    pub dynamic_entries: Vec<ImString>,
+}
+
+/// Widget display mode. Determines when the widget should be expanded.
+#[derive(serde::Deserialize, Debug, Clone, Copy)]
+pub enum Display {
+    /// The widget should always be in its expanded mode.
+    Always,
+    /// The widget should only be in its expanded mode when it has non-default value.
+    #[serde(rename = "When_Modified")]
+    WhenModified,
+    /// The widget should only be in its expanded mode whe the whole node is expanded.
+    #[serde(rename = "Expanded_Only")]
+    ExpandedOnly,
+}
+
+impl Default for Display {
+    fn default() -> Self {
+        Self::Always
+    }
+}
+
+/// The data of node port that this widget is attached to. Available immediately after widget
+/// creation. Can be updated later when the node data changes.
+#[derive(Debug, Clone, Default)]
+pub struct NodeData {
+    /// Argument info of the node port that this widget is attached to.
+    pub argument_info: span_tree::ArgumentInfo,
+    /// The current height of the node that the widget can use for layout.
+    /// TODO [pg]: remove and use the automatic layout instead.
+    pub node_height:   f32,
+}
+
+
+
+/// ==================
+/// === SampledFrp ===
+/// ==================
+
+/// Sampled version of widget FRP endpoints that can be used by lazily initialized widgets.
+/// Without sampling, the widget that is initialized after the endpoints are set would not receive
+/// the correct initial value.
+#[derive(Debug, Clone, CloneRef)]
+struct SampledFrp {
+    set_current_value: frp::Sampler<Option<ImString>>,
+    set_visible:       frp::Sampler<bool>,
+    set_focused:       frp::Sampler<bool>,
+    out_value_changed: frp::Any<Option<ImString>>,
+}
+
+
+
+// ==============
+// === Widget ===
+// ==============
+
+/// The node widget UI widget.
```

Comment:
`The node widget UI widget.` - sounds like magic. Also, TBH, I did not realy understood by reading it what exactly is it.